### DISCUSSION TITLE
Is there any way to call middleware from the client

### DISCUSSION THREAD

**Discussion by andriiradkevych** [vercel/next.js#38474] [Help]
**_middleware.ts**

```
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';

const isMobile = (userAgent: string) =>
  /iPhone|iPad|iPod|Android/i.test(userAgent);

export function middleware(req: NextRequest) {
  const userAgent = req.headers.get('user-agent');
  const { pathname, origin } = req.nextUrl;

  if (
    userAgent &&
    !pathname.includes('favicon.ico') &&
    !pathname.startsWith('/api') &&
    !req.headers.get('x-prerender-revalidate')
  ) {
    if (isMobile(userAgent)) {
      return NextResponse.rewrite(`${origin}/mobile${pathname}`);
    } else {
      return NextResponse.rewrite(`${origin}/desktop${pathname}`);
    }
  }

  return NextResponse.next();
}

````

```
    "next": "12.1.2",
    "react": "17.0.2",
```

Project structure of my project:

pages:
  - mobile
  - desktop
  
 
I've divided my pages on `mobile` and `desktop` version . I've made `middleware` , which can detect from which device user comes , and make `rewrite` depend on it (move user to version of my app depend on user device) . And solution seems ok for most of the cases , but what if user switch from `mobile` to `desktop` from `<1280` to `>1280` , imagine user on tablet and on horisontal device position it is more then `1280px`, but nothing will happens , it only make `rewrite` on page reload ( because there is middleware call happens) . Possible solution which I see , is to reload the page  , but it doesn't seems well , and I want to understand is there any way to make rewrite effect from client , or it is impossible to solve without page reload ?

**Comment by icyJoseph** [comment]
In your setup, the concept of mobile and desktop is based off the user agent. Unless users have their dev tools open, and switch to responsive mode, or directly mess with the user agent string, how could they switch between each?

I don't know if this is an edge case that you should put time into. Anyone who wants to monkey around with the user agent, or request desktop pages on mobile will game this. For example, in Chrome if you use responsive mode, then Chrome changes the user agent, but Firefox won't. And in general, spoofing the user agent is not very complicated.

Also, I know, often the reason to do this mobile/desktop content division is not necessarily for responsiveness, but because the actual value proposition of the site needs to be presented in a way different way, but CSS can really take us a long way, and help us to avoid this kind of situations.


### A duck-tape attempt

> Most likely should not be done, given the above considerations, and perhaps it won't even work when all things are considered

`router.replace` could've been used, if the pages use `getServerSideProps`, otherwise, it works only once. Once you detected a change between mobile and desktop, then you'd call `router.replace(router.asPath)`, and it'd SEEM to work, but once you build the application, it'll work only once.

> Again, this may work, but it is for sure a hack

However, if we add a random seed to the request made by `router.replace`:

```tsx
import type { NextPage } from "next";
import { useRouter } from "next/router";

// using the same for `Mobile`
const mode = "Desktop"; // or "Mobile"
const Desktop: NextPage = () => {
  const router = useRouter();

  const [path, query] = router.asPath.split("?");

  const seed = Math.random().toString(36).slice(2, 10);
  const triggerPath = `${path}${
    query ? `${query}&seed=${seed}` : `?seed=${seed}`
  }`;

  return (
    <div>
      {mode}
      <button onClick={() => router.replace(triggerPath, router.asPath)}>
        replace
      </button>
    </div>
  );
};

export default Desktop;
```

Then, if we have this `_app`:

```tsx
import type { AppProps } from "next/app";
import { useState } from "react";

function MyApp({ Component, pageProps }: AppProps) {
  const [count, setCount] = useState(0);
  return (
    <>
      <button onClick={() => setCount((x) => x + 1)}>Count: {count}</button>
      <Component {...pageProps} />
    </>
  );
}

export default MyApp;
```

And the same middleware you have, then navigation to `localhost:3000` in production build, yields `desktop`, in Chrome. 

If I increment up the counter, and activate `responsive` mode from the Chrome DevTools, and now click on replace, I am switched to the `mobile` view, but the counter remains.

I increment the counter again, and switch off `responsive` mode in Chrome DevTools, and click replace, then I see the `desktop` version, with the correct counter.

Then you would just need to trigger the replace callback automatically by listening to media query changes with JavaScript.

### useMediaQuery

To listen to media query changes the `useMediaQuery` hook is very good, for instance this implementation:

```ts
export function useMediaQuery(query: string, initial = false) {
    const [matches, setMatches] = useState(initial);

    useEffect(() => {
        const mediaQuery = window.matchMedia(query);

        const handler = () => setMatches(mediaQuery.matches);

        handler();

        if ('addEventListener' in mediaQuery) {
            mediaQuery.addEventListener('change', handler);

            return () => mediaQuery.removeEventListener('change', handler);
        }

        /**
         * This is documented as deprecated, but quite a lot of MacOS and iOS devices
         * do not implement it `addEventListener` yet!
         */
        mediaQuery.addListener(handler);

        return () => mediaQuery.removeListener(handler);
    }, [query]);

    useDebugValue(`${query} => ${matches}`);

    return matches;
}
```

**Reply by andriiradkevych** [reply]
Just for case like when someone use devtools , and play with it , offcourse I will not cover it , but I'm currious , is there any case when tablet in horizontal position doesn't have more then 1280px , I've found that tablet can be 1200px , this case , it can be a problem , for devices like tablets , but hope here is not a lot of cases like so ( o any of this ) @icyJoseph

**Reply by icyJoseph** [reply]
Right, but the user agent doesn't change, or does it? Because your check is based off the user agent.

You could still listen to on resize events, and try the router.replace call on the same route, with the random seed hidden, but that'll just arrive with the same user agent to the middleware, right?

**Reply by andriiradkevych** [reply]
God, wasn't thinking about it , for sure it is the same , and even if I would have it (rewrite from client) , it will not changed anything , thanks man
